### PR TITLE
Fix truncated small-file downloads

### PR DESCRIPTION
**Pull request by iwen-conf**
## Summary
- reject short but otherwise clean single-file responses instead of truncating and succeeding
- fail concurrent range downloads when a worker finishes short, cancel sibling workers, and verify total bytes before completion
- keep runtime status as downloading until finalization really succeeds, and tighten rename fallback handling during finalize

## Testing
- go test ./internal/engine/single ./internal/engine/concurrent ./internal/download ./internal/processing ./internal/core ./internal/engine/types -count=1 -timeout 180s

<h3>Greptile Summary</h3>

This PR hardens both the single-file and concurrent downloaders against truncated responses: small files that receive a short-but-clean HTTP body are now rejected with an explicit error instead of being silently saved, and concurrent range workers that finish short propagate a fatal `errIncompleteRange` disposition that cancels all sibling workers and aborts the download.

Key changes:
- **Single downloader** (`internal/engine/single/downloader.go`): replaces silent truncation with an explicit "incomplete download" error when `written != fileSize`.
- **Concurrent downloader** (`internal/engine/concurrent/`): replaces the flat `workerErrors` channel with a typed `workerResult` channel carrying a `RetryDisposition`. Fatal protocol errors (`errIncompleteRange`, `errRangeProtocol`) bypass the mirror-failover retry loop and immediately abort the download; transient errors are re-queued through a new `RetryTracker` that enforces a per-lineage stall budget before issuing a global cancel.
- **`verifiedBytes` aggregate check**: after all workers finish, a post-hoc byte total is verified against `fileSize`. When `d.State` is non-nil the check uses the capped `VerifiedProgress` counter; when nil it falls back to a raw `verifiedBytes` atomic (the over-count risk with hedging when state is nil was acknowledged in review).
- **`RuntimeStatus()`**: new method on `ProgressState` consolidates the pausing/paused/downloading status logic and intentionally omits the `Done` flag, keeping the status as "downloading" until lifecycle finalization truly completes (rename + DB write).
- **`finalizeCompletedFile`**: tightened to treat a pre-existing destination as a success *only* when the working `.surge` file has already been removed, preventing a silent false-success when the source file is still present.

Outstanding issues to watch:
- `DispositionDrop` is defined and handled in the balancer switch but is never generated by any worker code path — it is effectively dead code without documentation of its intended future use.
- `TestConcurrentDownloader_NilState` covers only the happy path; there is no test verifying that short-range rejection still works when `d.State == nil` and the fallback `verifiedBytes` path is in effect.

<h3>Confidence Score: 3/5</h3>

- Core logic is sound and well-tested for the primary paths, but several open issues in prior threads (stall-count off-by-one, `verifiedBytes` over-count under hedging with nil state, dead `result.err == nil` branch) and the new dead `DispositionDrop` value leave enough unresolved complexity to warrant another pass before merging.
- The fix correctly solves the stated problem — truncated small-file downloads are now rejected rather than silently saved — and the new test suite provides good breadth (short-range rejection, transient requeue recovery, stall-budget abort, completion-boundary requeue, nil-state). The architectural changes (typed dispositions, lineage tracking, aggregate byte verification) are well-structured. The score is lowered because: (1) `DispositionDrop` is dead code that could mislead future contributors into thinking a third result path exists; (2) the nil-state failure path for short-range responses has no test; (3) several issues raised in prior review threads remain open (stall count off-by-one, `verifiedBytes` over-count under hedging, dead `result.err == nil` branch, `pendingResults` counter semantics with `DispositionAbort`).
- `internal/engine/concurrent/runtime_types.go` (dead `DispositionDrop`), `internal/engine/concurrent/downloader.go` (open issues from prior threads: dead `result.err == nil` branch, `!=` vs `<` byte guard), `internal/engine/concurrent/concurrent_test.go` (missing nil-state failure-path test).

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| internal/engine/concurrent/runtime_types.go | New file introducing `queuedTask` envelope, `RetryTracker`, `pendingResultCounter`, and `workerResult` types. `DispositionDrop` is defined and handled in the balancer switch but is never generated by any worker code path — dead value with no documentation of its intended future use. |
| internal/engine/concurrent/downloader.go | Core changes: replaces flat `workerErrors` channel with `workerResult` channel carrying disposition, adds `retryTracker` for stall detection, adds `verifiedBytes` atomic as nil-state fallback, and moves the `downloadErr` check before the `context.Canceled` check. The `result.err == nil` branch is dead code (noted in prior threads); the `!=` equality guard on `verifiedBytes` is flagged in prior threads. |
| internal/engine/concurrent/worker.go | Worker now passes `verifiedBytes`, `workerResults`, and `pendingResults` references; reports fatal vs. requeue dispositions via `reportWorkerResult`; and detects short-range responses via `errIncompleteRange`. `isFatalTaskError` correctly bypasses the retry loop for protocol-level errors. |
| internal/engine/single/downloader.go | Removes the `preallocated` flag; now returns an explicit "incomplete download" error when `written != fileSize` instead of silently truncating the pre-allocated file. Clean and correct. |
| internal/processing/events.go | Tightens `finalizeCompletedFile`: rename failures now succeed (return nil) only when the working `.surge` file is already gone, preventing silent acceptance of a pre-existing destination file as a false-success when the working file is still present. |
| internal/engine/types/progress.go | New `RuntimeStatus()` method consolidates the pausing/paused/downloading status logic and deliberately omits the `Done` flag, ensuring the API reports "downloading" until lifecycle finalization completes. |
| internal/engine/concurrent/concurrent_test.go | Adds five new integration tests covering short-range rejection, transient requeue recovery, stall-budget abort, completion-boundary requeue, and nil-state download. The nil-state test (`TestConcurrentDownloader_NilState`) only validates the happy path, leaving the short-range-rejection fallback via `verifiedBytes` untested for nil state. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    W["Worker: downloadTask()"]
    W -->|"success: offset >= stopAt"| DONE["task complete — no result sent"]
    W -->|"errIncompleteRange / errRangeProtocol"| FATAL["isFatalTaskError → break retry loop"]
    W -->|"other error, retries exhausted"| RETRY["DispositionRequeue"]
    FATAL --> ABORT["reportWorkerResult(DispositionAbort)"]
    RETRY --> RWR["reportWorkerResult(DispositionRequeue)"]

    ABORT --> CH["workerResults channel"]
    RWR --> CH

    CH --> LOOP["Main loop: range workerResults"]
    LOOP -->|"disposition == Abort"| SETD["downloadErr = result.err\npendingResults.Complete()\ncancel()"]
    LOOP -->|"disposition == Requeue"| TRACKER["retryTracker.Evaluate(qTask)"]
    TRACKER -->|"DispositionRequeue"| REQUEUE["pendingResults.Requeue(queue, qTask)"]
    TRACKER -->|"DispositionAbort (stall budget)"| STALL["downloadErr = stalled error\ncancel()"]

    SETD --> END
    STALL --> END
    REQUEUE --> W

    END["After all workers finish"]
    END --> ERRCHECK{"downloadErr != nil?"}
    ERRCHECK -->|yes| RETURN_ERR["return downloadErr"]
    ERRCHECK -->|no| BYTECHECK{"fileSize > 0?"}
    BYTECHECK -->|yes| VERIFY["verified = loadVerifiedBytes(state, &verifiedBytes)\nif verified != fileSize → error"]
    BYTECHECK -->|no| FINALIZE["finalizeCompletedDownload()"]
    VERIFY -->|match| FINALIZE
    VERIFY -->|mismatch| RETURN_INCOMPLETE["return incomplete download error"]
```

<details open><summary><h3>Comments Outside Diff (2)</h3></summary>

1. `internal/engine/concurrent/downloader.go`, line 658-678 ([link](https://github.com/surge-downloader/surge/blob/9ca5c7dd077eb7ac7c57dd4c525ec767e9734359/internal/engine/concurrent/downloader.go#L658-L678)) 

   <a href="#">P0</a> **Hedged-worker failure causes false abort when primary already wrote all bytes**

   The `VerifiedProgress` guard is checked *after* `downloadErr` is returned, meaning a hedged worker's `errIncompleteRange` can fail the entire download even when the primary worker already wrote every byte to disk.

   Concrete scenario:
   1. Task `[0, 100 MB)` is live; an idle worker hedges it as `[50 MB, 100 MB)` with a shared `SharedMaxOffset`.
   2. The primary worker (W1) advances `SharedMaxOffset` to 100 MB and completes — `VerifiedProgress == fileSize`.
   3. The hedged worker (W2) gets an early EOF at 75 MB; `offset (75 MB) < StopAt (100 MB)` → returns `errIncompleteRange`.
   4. `DispositionAbort` sets `downloadErr` and calls `cancel()`.
   5. Line 660 returns `downloadErr` — download fails with "incomplete download" even though the file is 100% intact.

   The fix is to check aggregate verified bytes *before* returning any worker error, so that a hedged-task failure that occurs after all data was already written does not abort a good download:

   ```go
   // If all bytes are verified on disk, worker errors from racing hedged tasks can be
   // ignored — the primary already wrote the complete range.
   if d.State != nil && fileSize > 0 {
       if d.State.VerifiedProgress.Load() == fileSize {
           return finalizeCompletedDownload()
       }
   }

   if downloadErr != nil {
       return downloadErr
   }

   if downloadCtx.Err() == context.Canceled {
       return context.Canceled
   }

   if d.State != nil && fileSize > 0 {
       verified := d.State.VerifiedProgress.Load()
       if verified < fileSize {
           utils.Debug("concurrent download finalized short: verified=%d expected=%d url=%s", verified, fileSize, rawurl)
           return fmt.Errorf("incomplete download: got %d of %d bytes", verified, fileSize)
       }
   }

   return finalizeCompletedDownload()
   ```

   Note: The final guard uses `<` instead of `!=` because `UpdateChunkStatus` caps per-chunk increments via `remainingSpace`, so `VerifiedProgress` should never exceed `fileSize`; but using `!=` on an over-count would emit a misleading "incomplete download" error.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: internal/engine/concurrent/downloader.go
   Line: 658-678

   Comment:
   **Hedged-worker failure causes false abort when primary already wrote all bytes**

   The `VerifiedProgress` guard is checked *after* `downloadErr` is returned, meaning a hedged worker's `errIncompleteRange` can fail the entire download even when the primary worker already wrote every byte to disk.

   Concrete scenario:
   1. Task `[0, 100 MB)` is live; an idle worker hedges it as `[50 MB, 100 MB)` with a shared `SharedMaxOffset`.
   2. The primary worker (W1) advances `SharedMaxOffset` to 100 MB and completes — `VerifiedProgress == fileSize`.
   3. The hedged worker (W2) gets an early EOF at 75 MB; `offset (75 MB) < StopAt (100 MB)` → returns `errIncompleteRange`.
   4. `DispositionAbort` sets `downloadErr` and calls `cancel()`.
   5. Line 660 returns `downloadErr` — download fails with "incomplete download" even though the file is 100% intact.

   The fix is to check aggregate verified bytes *before* returning any worker error, so that a hedged-task failure that occurs after all data was already written does not abort a good download:

   ```go
   // If all bytes are verified on disk, worker errors from racing hedged tasks can be
   // ignored — the primary already wrote the complete range.
   if d.State != nil && fileSize > 0 {
       if d.State.VerifiedProgress.Load() == fileSize {
           return finalizeCompletedDownload()
       }
   }

   if downloadErr != nil {
       return downloadErr
   }

   if downloadCtx.Err() == context.Canceled {
       return context.Canceled
   }

   if d.State != nil && fileSize > 0 {
       verified := d.State.VerifiedProgress.Load()
       if verified < fileSize {
           utils.Debug("concurrent download finalized short: verified=%d expected=%d url=%s", verified, fileSize, rawurl)
           return fmt.Errorf("incomplete download: got %d of %d bytes", verified, fileSize)
       }
   }

   return finalizeCompletedDownload()
   ```

   Note: The final guard uses `<` instead of `!=` because `UpdateChunkStatus` caps per-chunk increments via `remainingSpace`, so `VerifiedProgress` should never exceed `fileSize`; but using `!=` on an over-count would emit a misleading "incomplete download" error.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

2. `internal/engine/concurrent/concurrent_test.go`, line 415-448 ([link](https://github.com/surge-downloader/surge/blob/ffa0d7a9acd7c78bbf5341fdcaeb6c54c497f9b3/internal/engine/concurrent/concurrent_test.go#L415-L448)) 

   <a href="#">P2</a> **Missing failure-path coverage for nil-state short-range rejection**

   `TestConcurrentDownloader_NilState` only exercises the happy path (full successful download). When `d.State == nil`, `loadVerifiedBytes` falls back to the `verifiedBytes` atomic counter, which is a different code path from the `d.State.VerifiedProgress` path exercised by every other integration test.

   There is no test confirming that a short-range server response is correctly rejected when `d.State == nil`. The per-task `errIncompleteRange` check in `downloadTask` would still fire (independent of state), but the aggregate `verifiedBytes` fallback check at finalization is exercised by zero tests in the failure direction.

   Consider adding a companion test alongside the existing `TestConcurrentDownloader_RejectsShortRangeResponse` that passes `nil` as the state argument to confirm both the per-task and aggregate rejection paths work in the nil-state configuration.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: internal/engine/concurrent/concurrent_test.go
   Line: 415-448

   Comment:
   **Missing failure-path coverage for nil-state short-range rejection**

   `TestConcurrentDownloader_NilState` only exercises the happy path (full successful download). When `d.State == nil`, `loadVerifiedBytes` falls back to the `verifiedBytes` atomic counter, which is a different code path from the `d.State.VerifiedProgress` path exercised by every other integration test.

   There is no test confirming that a short-range server response is correctly rejected when `d.State == nil`. The per-task `errIncompleteRange` check in `downloadTask` would still fire (independent of state), but the aggregate `verifiedBytes` fallback check at finalization is exercised by zero tests in the failure direction.

   Consider adding a companion test alongside the existing `TestConcurrentDownloader_RejectsShortRangeResponse` that passes `nil` as the state argument to confirm both the per-task and aggregate rejection paths work in the nil-state configuration.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: internal/engine/concurrent/runtime_types.go
Line: 23

Comment:
**`DispositionDrop` is dead code — never generated**

`DispositionDrop` is defined here and has a corresponding `case` in the balancer's result switch (in `downloader.go:588`), but no code path in `worker.go` ever sets `disposition = DispositionDrop`. The only caller of `reportWorkerResult` assigns either `DispositionRequeue` or `DispositionAbort`:

```go
// worker.go
disposition := DispositionRequeue
if isFatalTaskError(lastErr) {
    disposition = DispositionAbort
}
```

Because the channel is write-only from workers and `result.err` is never nil on this channel (workers only call `reportWorkerResult` inside `if lastErr != nil`), the `DispositionDrop` case is permanently unreachable. This creates a false impression that there is a third active disposition path, and a future maintainer may introduce a `pendingResults.Begin()` without a matching `Complete()` believing the `Drop` path provides a safe decrement.

If `DispositionDrop` is reserved for a future use-case, add a comment explaining the intent; otherwise remove the constant and the switch case to keep the contract clear.

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: internal/engine/concurrent/concurrent_test.go
Line: 415-448

Comment:
**Missing failure-path coverage for nil-state short-range rejection**

`TestConcurrentDownloader_NilState` only exercises the happy path (full successful download). When `d.State == nil`, `loadVerifiedBytes` falls back to the `verifiedBytes` atomic counter, which is a different code path from the `d.State.VerifiedProgress` path exercised by every other integration test.

There is no test confirming that a short-range server response is correctly rejected when `d.State == nil`. The per-task `errIncompleteRange` check in `downloadTask` would still fire (independent of state), but the aggregate `verifiedBytes` fallback check at finalization is exercised by zero tests in the failure direction.

Consider adding a companion test alongside the existing `TestConcurrentDownloader_RejectsShortRangeResponse` that passes `nil` as the state argument to confirm both the per-task and aggregate rejection paths work in the nil-state configuration.

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: internal/engine/concurrent/downloader.go
Line: 651-655

Comment:
**Resume path double-initializes `verifiedBytes` for fresh downloads**

For the fresh-download branch (non-resume), `verifiedBytes` is explicitly reset to 0 at line 523:

```go
verifiedBytes.Store(0)
```

But the variable is already zero-initialized at declaration (`var verifiedBytes atomic.Int64`). The explicit `.Store(0)` is redundant for fresh downloads, though it is necessary for the resume path (where it is first set to `savedState.Downloaded` at line 515, then correctly reset to 0 here if the resume check fails). A brief comment clarifying the intent — "reset in case resume state was partially applied before we decided not to resume" — would make the ordering dependency explicit and satisfy the why-not-what comment guideline.

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Last reviewed commit: ffa0d7a</sub>

**Context used:**

- Rule used - What: All code changes must include tests for edge... ([source](https://app.greptile.com/review/custom-context?memory=2b22782d-3452-4d55-b059-e631b2540ce8))
- Rule used - What: Comments must explain *why* code exists, not... ([source](https://app.greptile.com/review/custom-context?memory=4e45ef13-32c2-4753-8060-a44ef767144d))

### CODE REVIEW

**Comment by SuperCoolPencil** [comment]
> **Task re-queue removal changes retry semantics for all errors**
> 
> Previously, when a task exhausted all per-task retries (`maxRetries`), it was pushed back onto the queue so another worker — on a potentially different mirror — could attempt it from scratch:
> 
> ```go
> // old code
> if lastErr != nil {
>     queue.Push(task)
>     ...
> }
> ```
> 
> Returning an error here instead affects **all** task-failure reasons, not just short-range responses. A task that fails its last retry due to a transient network blip (timeout, connection reset, DNS hiccup) will now cancel the entire download rather than being re-queued.
> 
> In the old model, re-queuing effectively made per-task retries unbounded (each re-queue reset the attempt counter). The new model caps retries strictly at `maxRetries`. For users on unstable connections where a specific byte range is temporarily unreachable, this is a silent regression: downloads that previously healed themselves will now hard-fail.


Hey @iwen-conf 

Thanks for the PR. This is one issue I was worried about! A single worker failing could mean anything from 429 (too many requests) to server error. It doesn't necessarily mean the whole download is bad.

Can you please address this?

**Comment by iwen-conf** [comment]
I addressed the retry-semantics concern in the latest push (`7798e31`).

What changed:
- workers no longer decide whole-download failure for ordinary task errors
- transient task failures are requeued through a downloader-level control plane
- confirmed integrity/protocol violations (for example truncated range responses) still abort the download immediately
- the final `VerifiedProgress` byte-count gate remains as the last integrity check
- repeated no-progress retries on the same logical task lineage now abort instead of looping forever

I also added coverage for:
- truncated `206` responses returning an explicit incomplete-range error
- transient failures recovering via requeue
- repeated no-progress failures aborting via the stall budget
- completion-boundary requeue not losing tasks

Local verification:
- `go build ./...`
- `go test ./... -count=1`
- `go run github.com/golangci/golangci-lint/cmd/golangci-lint@latest run`

**Comment by iwen-conf** [comment]
Follow-up fix in `9ca5c7d`:

- balanced `pendingResults` on the `DispositionAbort` path
- added a small regression test for the pending-result counter accounting

The earlier implementation canceled correctly, but the abort branch leaked one pending-result count. That was mostly masked by `cancel()` forcing the monitor down the `balancerCtx.Done()` path, but the accounting was still wrong. The latest push fixes that and I re-ran:
- `go build ./...`
- `go test ./... -count=1`
- `go run github.com/golangci/golangci-lint/cmd/golangci-lint@latest run`

**Comment by iwen-conf** [comment]
Follow-up in `645a77d`:

- added a downloader-local verified-byte counter so the concurrent final byte-count gate also applies when `State` is nil
- kept the state-backed path for UI/progress, but removed the previous dependency on `State` existing for final integrity verification
- tightened the single downloader short-response guard to the direct `fileSize > 0 && written != fileSize` condition
- added coverage for the verified-byte fallback selection

Re-ran locally:
- `go build ./...`
- `go test ./... -count=1`
- `go run github.com/golangci/golangci-lint/cmd/golangci-lint@latest run`

**Comment by iwen-conf** [comment]
One broader observation from working through this PR: part of the churn here came from a few design boundaries in the downloader that feel more implicit than explicit today.

I think the behavior is in a much better place now, but I wanted to surface the architectural questions this exposed rather than keep making local fixes without checking the intended direction.

The places that felt under-specified to me were:
- worker vs downloader responsibilities: whether a worker should ever be able to implicitly terminate the whole download, or whether it should only report task outcomes and let the downloader own all cancel/requeue policy
- runtime scheduling metadata vs persisted task state: pause/resume tasks are persisted, but retry lineage / hedge / steal bookkeeping is runtime-only; I kept those separate in this PR, but I want to confirm that separation matches the intended model
- integrity accounting vs UI state: some correctness checks were effectively tied to `ProgressState` existing; I moved final verification toward downloader-local accounting, but I think it would help to make the rule explicit about which integrity checks may depend on state/UI objects and which may not
- completion semantics: once requeue / hedge / steal are in play, `queue.Len() == 0` is not enough to mean "done"; completion depends on queue state, active workers, and pending control-plane results together
- retry semantics as a policy: per-worker retries, task requeue, mirror failover, health-cancel, and no-progress aborts are closely related, but they are not yet expressed as one explicit policy boundary

I am not trying to broaden this PR into a redesign, but I would appreciate a deeper discussion on those boundaries. If the direction in this patch is roughly where you want the downloader to go, I am happy to follow up with a smaller cleanup/design PR after this one. If not, guidance on the intended control-plane boundaries would be very helpful so I do not keep solving the right bug with the wrong abstraction.

**Comment by SuperCoolPencil** [comment]
Hey @iwen-conf 

You're right, the archictecture needs to be refined a bit better. 

The ideal architecture would be something like this:
- Application layer: responsible for managing global state
- Processing later: responsible for a single download - ideally the one in control of all pause/resume and deletion/error handling
- Engine layer: responsible for just downloading - owning the worker pool - managing speeds (in future)
- Worker layer: should not be able to cancel the download, but can request a cancellation, 

 My current intent/direction:

  - Worker vs downloader: workers should never terminate the whole download implicitly. They report outcomes + signals; the downloader owns cancel/requeue/stop policy.
  - Runtime scheduling vs persisted state: keep retry/hedge/steal as runtime bookkeeping; persisted state should reflect resumable task progress only. That separation sounds right.
  - Integrity vs UI state: integrity/accounting should not depend on ProgressState or UI objects. UI can observe but must not gate correctness.
  - Completion semantics: agree queue.Len()==0 is insufficient. Completion should be derived from queue + in-flight + pending control-plane outcomes.
  - Retry semantics: we should make an explicit policy boundary that defines how per-worker retries, task requeue, mirror failover, health-cancel, and no-progress aborts compose.

  Your direction in this patch sounds aligned with that intent. Please keep this PR focused, and I’d definitely welcome a follow-up cleanup/design PR to codify the boundaries and invariants (maybe a short policy doc + refactor).

File: internal/engine/concurrent/downloader.go [hunk 1]
```diff
@@ -593,12 +660,20 @@ func (d *ConcurrentDownloader) Download(ctx context.Context, rawurl string, cand
...
 
-	if downloadErr != nil {
-		return downloadErr
+	if fileSize > 0 {
+		verified := loadVerifiedBytes(d.State, &verifiedBytes)
+		if verified != fileSize {
+			utils.Debug("concurrent download finalized short: verified=%d expected=%d url=%s", verified, fileSize, rawurl)
+			return fmt.Errorf("incomplete download: got %d of %d bytes", verified, fileSize)
+		}
 	}
```
[the remark this answers is not in the document]
Comment (SuperCoolPencil) [reply internal/engine/concurrent/downloader.go:677]: This is a valid issue, and I am afraid I don't know an easy way to fix this right now. 

HedgeWork will be a problem when trying to increment VerifiedProgress

**Comment by SuperCoolPencil** [comment]
Everything seems stable, the HedgeWork bug shouldnt affect us as we always overcount and never undercount so the download will be marked as success. Are we ready to merge?

**Comment by iwen-conf** [comment]
Thanks — I agree the remaining HedgeWork / verified-progress interaction is a deeper control-plane issue than I want to solve opportunistically in this bugfix PR.

I think this PR is in a good place to stay focused on the truncated-download / retry-semantics fix. If I arrive at a cleaner architecture for HedgeWork and progress accounting later, I will follow up with a separate PR rather than keep stretching this one.

If it is easier to discuss that design direction directly at some point, you can also reach me by email at `iluwenconf@gmail.com`.

**Comment by SuperCoolPencil** [comment]
So just before I merge, I questions from an architectural standpoint. 

I'm trying to wrap my head around the exact edge case this solves. 

Since we already have controls in place for both local write failures and standard HTTP errors, in what real-world scenarios are we seeing the verified progress end up being less than the expected file size?

Just want to make sure I fully understand the root bug this fixes
